falcon/routing/compiled.py

```diff
@@ -531,6 +568,23 @@ def _instantiate_converter(self, klass, argstr=None):
         src = '{0}({1})'.format(klass.__name__, argstr)
         return eval(src, {klass.__name__: klass})
 
+    def _compile_and_find(self, path, _return_values, _patterns, _converters, params):
+        """Compiles the router, sets the `_find` attribute and returns its result
+
+        This method is set to the `_find` attribute to delay the compilation of the
+        router until its used for the first time. subsequent calls to `_find` will
+        be processed by the actual routing function.
+        This method must have the same signature as the function returned by the
+        :meth:`.CompiledRouter._compile`.
+        """
+        # NOTE(caselit): replace the find with the actual method
+        self._find = self._compile()
```

Comment:
should we have a lock of some kind here to support multi-threading/async invocations?

Reply:
I don't think we need a lock here, assuming `_compile()` itself is threadsafe. Inside `_compile()`, the only potential issue would be one thread mutating the set of routes while another thread is attempting to generate the AST. But there aren't any points in there where the GIL or a green thread or coroutine might yield execution, so we should be safe.

Reply:
Reposting my somewhat similar comments from Gitter

> @CaselIT you could probably rely on atomicity of Python structures guarded by the GIL, but then it is IMHO important to be crystal clear about that in the comments and/or docs.

> E.g., setting a dictionary or object key/value is atomic under the GIL. So just compiling and replacing the routing method should work fine from that perspective.

> One aspect that might be worth considering in this approach is, however, an API facing high load directly when from the second 0. That may trigger a large number of parallel compilations. However, this is only an issue when using threads, and not async/green threads, and APIs are normally brought online by means of an LB check or similar, that could be used as an opportunity to compile.

Reply:
I'm not sure if compile is thread safe. I mean, repeated invocations should produce the same output, but since it mutates the state of the router (it sets return value and patters) it may happen that an invocation of find fails/returna the wrong result since it happens with the wrong values?

Reply:
I guess a simple
```py
with self._compile_lock:
    if self._find == self._compile_and_find:
        self._find = self._compile()
```
would not have a lot of downsides and we avoid multiple compilation in multi threading envs